### PR TITLE
usb: fix unaligned access to interface and device descriptor

### PR DESCRIPTION
**Pull request by jfischer-no**
Fix unaligned access to interface and device descriptor.

see https://github.com/zephyrproject-rtos/zephyr/pull/4661#pullrequestreview-130210406

### CODE REVIEW

**Comment by aurel32** [comment]
Hmm the structures are correctly marked as __packed, and the access is not done through a pointer of a different type, so GCC should generate unaligned access code itself. That looks like a compiler bug to me, that said, the workaround is correct.

File: subsys/usb/class/hid/core.c [hunk 1]
```diff
@@ -68,8 +68,8 @@ USBD_CLASS_DESCR_DEFINE(primary) struct usb_hid_config hid_cfg = {
 
 static void usb_set_hid_report_size(u16_t report_desc_size)
 {
-	hid_cfg.if0_hid.subdesc[0].wDescriptorLength =
-		sys_cpu_to_le16(report_desc_size);
+	UNALIGNED_PUT(sys_cpu_to_le16(report_desc_size),
+		      &hid_cfg.if0_hid.subdesc[0].wDescriptorLength);
 }
```
Comment (finikorg) [inline subsys/usb/class/hid/core.c]: I wonder could we use here sys_put_le16()

File: subsys/usb/class/hid/core.c [hunk 2]
```diff
@@ -211,7 +211,8 @@ static struct usb_ep_cfg_data hid_ep_data[] = {
 
 static void hid_interface_config(u8_t bInterfaceNumber)
 {
-	hid_cfg.if0.bInterfaceNumber = bInterfaceNumber;
+	UNALIGNED_PUT(bInterfaceNumber,
+		      &hid_cfg.if0.bInterfaceNumber);
 }
```
Comment (finikorg) [inline subsys/usb/class/hid/core.c]: It does not need alignment.

File: subsys/usb/class/netusb/netusb.c [hunk 3]
```diff
@@ -367,29 +367,42 @@ static void netusb_interface_config(u8_t bInterfaceNumber)
...
+		      &cdc_ecm_cfg.if0.bInterfaceNumber);
+	UNALIGNED_PUT(bInterfaceNumber,
+		      &cdc_ecm_cfg.if0_union.bControlInterface);
+	UNALIGNED_PUT((bInterfaceNumber + 1),
+		      &cdc_ecm_cfg.if0_union.bSubordinateInterface0);
+	UNALIGNED_PUT((bInterfaceNumber + 1),
+		      &cdc_ecm_cfg.if1_0.bInterfaceNumber);
+	UNALIGNED_PUT((bInterfaceNumber + 1),
+		      &cdc_ecm_cfg.if1_1.bInterfaceNumber);
 #ifdef CONFIG_USB_COMPOSITE_DEVICE
```
Comment (finikorg) [inline subsys/usb/class/netusb/netusb.c]: same here

**Review by finikorg — CHANGES_REQUESTED** [review]

File: subsys/usb/class/hid/core.c (hunk 2 above)
Comment (jfischer-phytec-iot) [reply subsys/usb/class/hid/core.c]: You might be right, I have no knowledge about how Intel S1000 supports unaligned accesses

File: subsys/usb/class/hid/core.c (hunk 1 above)
Comment (jfischer-phytec-iot) [reply subsys/usb/class/hid/core.c]: I see it now, should work, @SavinayDharmappa can you try it?

**Comment by jfischer-phytec-iot** [comment]
revised in #11619

**Pull request closed without merging**